lib/client/conntest/database.go

```diff
@@ -0,0 +1,394 @@
+		ResourceType:        types.KindDatabaseServer,
+		Limit:               1,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	databaseServers, err := types.ResourcesWithLabels(listResourcesResponse.Resources).AsDatabaseServers()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if len(databaseServers) == 0 {
+		connDiag, err := s.appendDiagnosticTrace(ctx,
+			connectionDiagnosticID,
+			types.ConnectionDiagnosticTrace_RBAC_DATABASE,
+			"You are not authorized to access this Database. "+
+				"Ensure your role grants access by adding it to the 'db_labels' property. "+
+				"This can also happen when you don't have a Database Agent proxying the database - "+
+				"you can fix that by adding the database labels to the 'db_service.resources.labels' in 'teleport.yaml' file of the database agent.",
+			trace.NotFound("%s not found", req.ResourceName),
+		)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		return connDiag, nil
+	}
+
+	databaseServer := databaseServers[0]
+	databaseProtocol := databaseServer.GetDatabase().GetProtocol()
+	databaseServiceName := databaseServer.GetName()
+
+	var databasePinger DatabasePinger
+
+	switch databaseProtocol {
+	case defaults.ProtocolPostgres:
+		databasePinger = database.PostgresPinger{}
+	default:
+		return nil, trace.Errorf("Database protocol not supported yet for Test Connection")
+	}
+
+	if _, err := s.appendDiagnosticTrace(ctx,
+		connectionDiagnosticID,
+		types.ConnectionDiagnosticTrace_RBAC_DATABASE,
+		"A Database Agent is available to proxy the connection to the Database.",
+		nil,
+	); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	list, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	proxyMiddleware := client.NewALPNCertChecker().
+		WithIdentityCheckerFunc(func(identity *tlsca.Identity) error {
+			if req.DatabaseUser != identity.RouteToDatabase.Username {
+				msg := fmt.Sprintf("certificate subject is for user %s, but need %s", identity.RouteToDatabase.Username, req.DatabaseUser)
+				return trace.Wrap(errors.New(msg))
+			}
+
+			if req.DatabaseName != identity.RouteToDatabase.Database {
+				msg := fmt.Sprintf("certificate subject is for database name %s, but need %s", identity.RouteToDatabase.Database, req.DatabaseName)
+				return trace.Wrap(errors.New(msg))
+			}
+
+			return nil
+		})
+
+	err = client.RunALPNAuthTunnel(ctx, client.RunALPNAuthTunnelRequest{
+		Client:                 s.cfg.UserClient,
+		Listener:               list,
+		Protocol:               databaseProtocol,
+		WebProxyAddr:           s.webProxyAddr,
+		ProxyMiddleware:        proxyMiddleware,
+		ConnectionDiagnosticID: connectionDiagnosticID,
+		RouteToDatabase: proto.RouteToDatabase{
+			Protocol:    databaseProtocol,
+			Username:    req.DatabaseUser,
+			Database:    req.DatabaseName,
+			ServiceName: databaseServiceName,
+		},
+		Insecure: true, // TODO(marco): remove this
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```

Comment:
We we need to call the web proxy again with and setup the local reverse tunnel ? 

The `DatabaseConnectionTester) TestConnection` call happens on the Teleport Proxy side so there is not need to call  ALPN PROY that dials the Teleport Proxy because the flow is already in Teleport Proxy and all and we have already all information needed to dial database so in this scenario we only need to dial the db agent over a reverse tunnel and that should be quite easy. 

ALPN Proxy was intended to run on the client side and changing this and adding the local proxy server logic in Teleport Proxy flow will make the code harder to maintain. 

Does it makes sense ?

Reply:
It does, yes. Thank you.

The goal of Test Connection is to be as closer to a real scenario as possible.
The user will use `tsh` to create a tunnel and then use their `psql` client to connect to it.

Creating this extra loop ensures we are closer to that real scenario.

I'm fine removing it - I mean, we are not using `tsh` anyway so we are already far from a real use case.

Reply:
We ended up deciding on not using the proxy's reverse tunnel from the Database Service.